File: 프로그래머스/2/258711. 도넛과 막대 그래프/도넛과 막대 그래프.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

vector<int> solution(vector<vector<int>> edges) {
    int max_node = 0;
    for (auto e : edges) {
        max_node = max(max(max_node, e[0]), e[1]);
    }
    
    vector<int> indegree(max_node+1, 0), outdegree(max_node+1, 0);
    for (auto e : edges) {
        outdegree[e[0]]++;
        indegree[e[1]]++;
    }
    
    int a = 0, b = 0, c = 0, d = 0;
    for (int i = 1; i <= max_node; i++) {
        if (indegree[i] == 0 && outdegree[i] >= 2) a = i;
        else if (outdegree[i] == 0 && indegree[i] >= 1) c++;
        else if (indegree[i] >= 2 && outdegree[i] >= 2) d++;
    }
    b = outdegree[a] - c - d;
    return {a, b, c, d};
}

File: 프로그래머스/2/258711. 도넛과 막대 그래프/도넛과 막대 그래프.java
import java.util.*;
class Solution {
    static int MAX = 1_000_001;
    public int[] solution(int[][] edges) {
        int[] answer = new int[4];
        
        int[] indegree = new int[MAX];
        int[] outdegree = new int[MAX];
        int maxNumber = 0;
        
        for (int[] edge : edges) {
            indegree[edge[1]]++;
            outdegree[edge[0]]++;
            maxNumber = Math.max(maxNumber, edge[0]);
            maxNumber = Math.max(maxNumber, edge[1]);
        }
        
        // 정점
        for (int i = 1; i <= maxNumber; i++) {
            if (indegree[i] == 0 && outdegree[i] >= 2) answer[0] = i;
        }
        
        // 막대 모양
        for (int i = 1; i <= maxNumber; i++) {
            if (outdegree[i] == 0 && indegree[i] >= 1) answer[2]++;
        }
        
        // 8자 모양
        for (int i = 1; i <= maxNumber; i++) {
            if (outdegree[i] >= 2 && indegree[i] >= 2) answer[3]++;
        }
        
        // 도넛 모양
        answer[1] = outdegree[answer[0]] - answer[2] - answer[3];
        
        return answer;
    }
}